microsoft/DevHomeAzureExtension
# Toast notifications for Azure pull requests (Rejected and Approved)

## Summary of the pull request
Adds toast notifications for when a recent developer's pull request is either Rejected by policy or passes all policy and is Approved.

![image](https://github.com/user-attachments/assets/c115443c-1acb-4054-b88e-20e092519acd)

Also adds Pull Request policy status to the Pull Request data for all pull requests, which can be utilized by Pull Request widget for displaying status icons. The display status change is not part of this PR, but this PR enables it.
  
## References and relevant issues
#111 

## Detailed description of the pull request / Additional comments
Changes:
* Added new DataModel objects for PullRequestPolicyStatus and Notifications.
* Updated pull request fetching code to utilize Repository information.
* Added updater for checking for developer pull request changes. This relies on the Repository cache being updated to discover the developer's active repositories and can be delayed the first time by several minutes.
* Added Notification handling and pushing after updating the developer's pull requests. Notification code and display logic is very similar to that of the GitHub extension and has similar behavior.
* Updated and refactored Pull Request fetching behavior to be easier to read and process.
* Updated the data model for Pull Request to use Id for Identity and WorkItemType, which reduces the record size significantly.
* Added data manager interface methods to fetch WorkItemType and Identity by Id.
* Updated Widgets to have access to the new Pull Request Policy status.
* Added tests for Notifications and PullRequestStatus objects.

Notes:
* Only a developer's own pull requests will receive notifications. This is for developer accounts the user has added via the extension.
* Pull Requests older than 14 days are considered not notification worthy so as to prevent notification spam of ancient pull requests. This is an easily changeable constant should we desire more/less retention.
* A pull request's update time is considered to be the most recent commit push or its creation date, whichever is newer. Pull requests in active development will have notifications (as long as it's been pushed within the last 14 days). Pull requests that have been sitting around for longer than that will not have notifications.
* Only notifications for policy evaluation failures or approvals are created. For example, a pull request that has had a rejection from a required reviewer will display a notification, or one that fails a build policy check. A pull request that is running a policy or waiting for approval will not give notifications unless some other policy check fails.

## Validation steps performed
* Verified widgets still work as expected.
* All Data Model tests passed, including the new one for notifications and pull request policy status.
* Lots of manual verification with own pull request and test pull requests that the data retention and notification display behavior is correct.

## PR checklist
- [x] Closes #111
